test/onnx/onnx_test_common.py

```diff
@@ -459,14 +459,24 @@ def _compare_pytorch_onnx_with_ort(
     else:
         raise ValueError(f"Unsupported model type: {model_type}")
 
-    if len(ref_outputs) != len(ort_outputs):
-        raise AssertionError(
-            f"Expected {len(ref_outputs)} outputs, got {len(ort_outputs)}"
-        )
-    for ref_output, ort_output in zip(ref_outputs, ort_outputs):
-        torch.testing.assert_close(
-            ref_output, torch.tensor(ort_output), rtol=rtol, atol=atol
-        )
+    for output_format in (
+        "pytorch",
+        "onnx",
+    ):  # Test output match with pytorch and onnx output format
+        if output_format == "pytorch":
+            ort_outputs = onnx_program.adapt_onnx_outputs_to_torch(ort_outputs)
```

Comment:
I didn't do this because it felt non-trivial. onnx outputs are more organized. It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

Reply:
> I didn't do this because it felt non-trivial. onnx outputs are more organized. 
> It can only be tensors, or (nested) list of tensors. Whereas for torch, the output can be anything. So mapping back from onnx to torch is considerably harder (how to insert back the Nones at the correct location where we discarded? how to construct back the arbitrary torch model output classes/datastructures that we flattened?).

That is a good point. My PR doesn't include a test with a very complex output that need to be rebuilt. This PR only has a test which requires buffers to be prepended to the original output.

I am wondering whether `pytree` could do that for us since it records the spec of the original data?

> I see that the goal was rather to eliminate the extra mutated buffers. I feel finding the way to include those outputs, when they exist because we are exporting ExportProgram, in the `ref_outputs` and have them properly accuracy checked might be more desirable.

We already do that too :) `adapt_onnx_outputs_to_torch` is not strictly necessary to support mutated buffers, but because you had proposed in another [review](https://github.com/pytorch/pytorch/pull/109650#discussion_r1331940092) I took this opportunity to implement here and improve the UX

Maybe we can split this PR with the essentials for supporting mutated buffers and move the `adapt_onnx_outputs_to_torch` into a new change we might need to dig deeper